### ISSUE TITLE
Invoke-CMApplyDriverPackage.ps1 not using $FallBackSKU

### ISSUE THREAD

**Issue by CornHead764** [SCConfigMgr/ConfigMgr#202]
So I was attempting to use this script on a Dell Optiplex 7010, and the script kept failing. Looking at the logs, below is what it was detecting:

-Manufacturer determined as: Dell
-Computer model determined as: OptiPlex 7010
-Computer SKU determined as: OptiPlex 7010
-Fallback SKU determined as: 0577

After this, once it finds the 7010 driver package, we see this in the logs:

-Attempting to find a match for driver package: Drivers - Dell OptiPlex 7010 - Windows 10 x64
-Unable to match computer model using detection method: SystemSKU (OptiPlex 7010)
-Fallback from SystemSKU match found for computer model instead using detection method: SystemSKU
-Driver package does not meet computer model, manufacturer and operating system and architecture criteria: Drivers - Dell OptiPlex 7010 - Windows 10 x64

So we can see the script detects the SKU as a fallback, and the model as the SKU. I didn't dig into the SKU detection logic, but figured I'd look at the fallback, as if it's detecting that, it should still work. Looking at the script, there are only 3 references to the $FallBackSKU variable. and those are as follows:

Line 598: `$FallBackSKU = [regex]::Matches($OEMString, '\[\S*]')[0].Value.TrimStart("[").TrimEnd("]")`

Line 610-612:
        `if (-not ([string]::IsNullOrEmpty($FallBackSKU))) {
		Write-CMLogEntry -Value "Fallback SKU determined as: $FallBackSKU" -Severity 1
	}`

So really, the variable is detected, filled, and logged, but never used.

To make it work, I simply added a clause to the if statement on line 813, which now reads:
`if (($Package.PackageManufacturer -match $ComputerManufacturer) -and ($Package.PackageName -match $OSName) -and ($Package.PackageName -match $OSImageArchitecture) -and (($Package.PackageDescription -match $SystemSKU) -or ($Package.PackageDescription -match $FallBackSKU))) {`

The change is the statement after the last -and. Previously, it didn't reference the $FallBackSKU variable, but now that it does, if either the $SystemSKU or $FallBackSKU are in the package description (which was already the case with the 7010), then the statement will evaluate as true.

This worked as a fix at 4:45, so there may be a better way to fix this, but that's what I found that worked for my scenario. Please let me know if you need any additional information, validation, or testing.

**Comment by timothyd09** [comment]
Having this same issue and I simply commented out the validation "(($Package.PackageDescription -match $SystemSKU)" as it has already been evaluated in the previous section so I didn't feel it needed to be evaluated again.

**Comment by celadyn** [comment]
Having the same issue with a Dell Optiplex 7010. Both model and SKU (0577) detection fail and no drivers are installed. About to dive into the script and make some change similar to the above commenters :) wonder why it's just this one that fails? No other model appears to generate the same behavior (that I've seen).

```
Manufacturer determined as: Dell
Computer model determined as: OptiPlex 7010
Computer SKU determined as: OptiPlex 7010
Fallback SKU determined as: 0577
...
Attempting to find a match for driver package: Drivers - Dell Optiplex 7010 - Windows 10 x64   (CCM00C03)
Unable to match computer model using detection method: SystemSKU (OptiPlex 7010)
Fallback from SystemSKU match found for computer model instead using detection method: SystemSKU (OptiPlex 7010)
Driver package does not meet computer model, manufacturer and operating system and architecture criteria: Drivers - Dell Optiplex 7010 - Windows 10 x64 (CCM00C03)
```

**Comment by NickolajA** [comment]
Could you please try the 3.0.0 version of the script to see if the UseDriverFallback parameter and functionality works better for you?

https://github.com/SCConfigMgr/ConfigMgr/blob/master/Operating%20System%20Deployment/Drivers/Invoke-CMApplyDriverPackage_3.0.0_Preview.ps1

**Comment by NickolajA** [comment]
Any update?

**Comment by shaylesh** [comment]
Hi Nickolaj,

Is the Dell Optiplex 7010 issue fixed in version 3.0.0 ? I am facing exactly same issue.

**Comment by NickolajA** [comment]
Use the new improved DebugMode where you can test if any model will match the correct driver package. An example is on our docs page on the blog. I can't remember this specific issue though.

**Comment by agrooms** [comment]
The Dell Optiplex 3020M exhibits the same behavior

**Comment by NickolajA** [comment]
Could you share the full ApplyDriverPackage.log from that model?

**Comment by agrooms** [comment]
Attached

Allen L. Grooms | IS Network Analyst
Information Systems | Great River Health
1221 South Gear Ave. | West Burlington, IA 52655
Office: 319-768-4466 | Department: 319-768-4400
www.greatrivermedical.org<http://www.greatrivermedical.org/>
www.Facebook.com/GreatRiverHealth<http://www.Facebook.com/GreatRiverHealth> | www.Twitter.com/GreatRiverMed<http://www.twitter.com/GreatRiverMed>

Information in this communication, including attachments, is confidential and intended only for the addressee(s). This communication may contain privileged, confidential, proprietary or trade secret information entitled to protection or exemption from disclosure under law. If you are not an intended recipient, please know that any use, distribution or copying of this communication, or any action taken based on the information in this communication, is unauthorized and may be unlawful. If you received this communication in error, please notify the sender and delete this communication from your device.

**Comment by agrooms** [comment]
[ApplyDriverPackage.zip](https://github.com/SCConfigMgr/ConfigMgr/files/4371027/ApplyDriverPackage.zip)

**Comment by agrooms** [comment]
I can confirm the same on the Dell Optiplex 7010 as reported earlier.  I can force the update again and pull that log if would help.

**Comment by NickolajA** [comment]
That's not using the latest version 3.0.0, it's using version 2.2.7. Could you test with the new, please? A lot of the logic has been re-written.

**Comment by agrooms** [comment]
Agreed, I am not on the beta/preview. I will download it and give it a try tomorrow. Sorry if my post was misleading to make you think I was using the 3.0 version.

Please forgive any grammar and typing issues as this response is being sent from a mobile device.

Allen L. Grooms
________________________________

**Comment by NickolajA** [comment]
No worries, it's actually not in Preview anymore. Was released this weekend!

**Comment by agrooms** [comment]
I can confirm that the new 3.0 version fixed the issue with the Dell Optiplex 3020M issue.  Will try and find a 7010 to test next.

Allen L. Grooms | IS Network Analyst
Information Systems | Great River Health
1221 South Gear Ave. | West Burlington, IA 52655
Office: 319-768-4466 | Department: 319-768-4400
www.greatrivermedical.org<http://www.greatrivermedical.org/>
www.Facebook.com/GreatRiverHealth<http://www.Facebook.com/GreatRiverHealth> | www.Twitter.com/GreatRiverMed<http://www.twitter.com/GreatRiverMed>

Information in this communication, including attachments, is confidential and intended only for the addressee(s). This communication may contain privileged, confidential, proprietary or trade secret information entitled to protection or exemption from disclosure under law. If you are not an intended recipient, please know that any use, distribution or copying of this communication, or any action taken based on the information in this communication, is unauthorized and may be unlawful. If you received this communication in error, please notify the sender and delete this communication from your device.

**Comment by NickolajA** [comment]
Great to hear! Looking forward to the results from the other model.

**Comment by agrooms** [comment]
I can confirm that the Dell OptiPlex 7010 is now working as well with the new Update Drivers script.

MANY thanks to you and your team for developing these AWESOME tools.

Allen L. Grooms | IS Network Analyst
Information Systems | Great River Health
1221 South Gear Ave. | West Burlington, IA 52655
Office: 319-768-4466 | Department: 319-768-4400
www.greatrivermedical.org<http://www.greatrivermedical.org/>
www.Facebook.com/GreatRiverHealth<http://www.Facebook.com/GreatRiverHealth> | www.Twitter.com/GreatRiverMed<http://www.twitter.com/GreatRiverMed>

Information in this communication, including attachments, is confidential and intended only for the addressee(s). This communication may contain privileged, confidential, proprietary or trade secret information entitled to protection or exemption from disclosure under law. If you are not an intended recipient, please know that any use, distribution or copying of this communication, or any action taken based on the information in this communication, is unauthorized and may be unlawful. If you received this communication in error, please notify the sender and delete this communication from your device.

**Comment by timothyd09** [comment]
Thanks for working on this bug.  I haven't had any time to test it with
everyone moving to remote work, but I will give it a shot when I have a
chance.

Regards,

*Tim De Haan*
*Domain Architect / SCCM Administrator*
*Operations Infrastructure & Applications*
*Health Sciences Technology (HST)*
*University of Minnesota*
*tdehaan@umn.edu <tdehaan@umn.edu>*
*612-301-1479*

**Comment by agrooms** [comment]
Sorry to be a pain in the butt.  I am expanding my testing and am finding that my HP 840 laptops are not working now.  The script seems to find the appropriate manufacturer and model and sees that there are two driver packages but then abort.  See attached log file.  Thanks in advance for your support and assistance.

**Comment by agrooms** [comment]
[ApplyDriverPackage.zip](https://github.com/SCConfigMgr/ConfigMgr/files/4375386/ApplyDriverPackage.zip)
